src/metabase/lib/extraction.cljc

```diff
@@ -0,0 +1,117 @@
+   (host-extraction      column)])
+
+(defn- url-extractions [column]
+  [(domain-extraction    column)
+   (subdomain-extraction column)
+   (host-extraction      column)])
+
+(mu/defn column-extractions :- [:maybe [:sequential ::lib.schema.extraction/extraction]]
+  "Column extractions are a set of transformations possible on a given `column`, based on its type.
+
+  For example, we might extract the day of the week from a temporal column, or the domain name from an email or URL.
+
+  Returns a list of possible column extractions for the given column, or `nil` if there are none."
+  [query  :- ::lib.schema/query
+   column :- ::lib.schema.metadata/column]
+  (cond
+    (lib.types.isa/temporal? column) (column-extract-temporal-units column)
+
+    ;; The URL and email extractions are powered by regular expressions, and not every database supports those.
+    ;; If the target database doesn't support :regex feature, return nil.
+    (not (regex-available? query))   nil
+    (lib.types.isa/email? column)    (email-extractions column)
+    (lib.types.isa/URL? column)      (url-extractions column)))
+
+(defmethod lib.metadata.calculation/display-info-method ::extraction
+  [_query _stage-number extraction]
+  (dissoc extraction :lib/type :column))
+
+(defn- case-expression
+  "Creates a case expression with a condition for each value of the unit."
+  [expression-fn unit n]
+  (lib.expression/case
+    (for [raw-value (range 1 (inc n))]
+      [(lib.filter/= (expression-fn) raw-value) (shared.ut/format-unit raw-value unit)])
+    ""))
+
+(defn- extraction-expression [column tag]
+  (case tag
+    ;; Temporal extractions
+    :hour-of-day     (lib.expression/get-hour column)
+    :day-of-month    (lib.expression/get-day column)
+    :day-of-week     (case-expression #(lib.expression/get-day-of-week column) tag 7)
+    :month-of-year   (case-expression #(lib.expression/get-month column) tag 12)
+    :quarter-of-year (case-expression #(lib.expression/get-quarter column) tag 4)
+    :year            (lib.expression/get-year column)
+    ;; URLs and emails
+    :domain          (lib.expression/domain column)
+    :subdomain       (lib.expression/subdomain column)
+    :host            (lib.expression/host column)))
+
+(mu/defn extract :- ::lib.schema/query
+  "Given a query, stage and extraction as returned by [[column-extractions]], apply that extraction to the query."
+  [query                :- ::lib.schema/query
+   stage-number         :- :int
+   {:keys [column display-name tag]} :- ::lib.schema.extraction/extraction]
+  ;; Currently this is very simple: use the `:tag` as an expression function and the column as the only argument.
+  (let [unique-name-fn         (lib.util/unique-name-generator)]
+    (doseq [col-name (->> (lib.util/query-stage query stage-number)
+                          (lib.metadata.calculation/returned-columns query stage-number)
+                          (map :name))]
+      (unique-name-fn col-name))
```

Comment:
This feels like a a generally useful util fn. Maybe even a `unique-name-generator` arity that "primes' it.

Reply:
Ooh, I like that. I've added this extra arity to `lib.util/unique-name-generator` and used it here.